pkg/apis/rollouts/validation/validation_test.go

```diff
@@ -129,6 +129,80 @@ func TestValidateRolloutStrategyBlueGreen(t *testing.T) {
 	assert.Equal(t, ScaleDownLimitLargerThanRevisionLimit, allErrs[1].Detail)
 }
 
+func TestValidateRolloutStrategyCanaryMissingServiceNames(t *testing.T) {
```

Comment:
Can we add a few test that use two traffic routers one where both are required to have canary service and one where it one requires it and one does not.

Reply:
Added some test cases with multiple routers where one requires the canary services and one or more do not to ensure they error as expected.